compiler/rustc_ast_passes/src/ast_validation.rs

```diff
@@ -1640,129 +1449,120 @@ fn deny_equality_constraints(
-                                            }
-                                            .into();
-                                        }
+        if let TyKind::Path(None, qself_ty_path) = &qself.ty.kind {
+            if let [PathSegment { ident: qself_ty_ident, args: None, .. }] =
+                &qself_ty_path.segments[..]
+            {
+                if generics.params.iter().any(|param| param.ident == *qself_ty_ident) {
+                    if let [PathSegment { ident: assoc_ty_ident, args: assoc_ty_args, .. }] =
+                        &full_path.segments[qself.position..]
+                    {
+                        // Build `<Bar = RhsTy>`.
+                        let new_arg = AngleBracketedArg::Constraint(AssocConstraint {
+                            id: rustc_ast::node_id::DUMMY_NODE_ID,
+                            ident: *assoc_ty_ident,
+                            gen_args: assoc_ty_args.as_ref().map(|p| (**p).clone()),
+                            kind: AssocConstraintKind::Equality {
+                                term: predicate.rhs_ty.clone().into(),
+                            },
+                            span: assoc_ty_ident.span,
+                        });
+
+                        let bound_opt = {
+                            // Make a new `Path` from `foo::Bar` to `Foo<Bar = RhsTy>`.
+                            let mut new_bound = full_path.clone();
+                            // Remove `Bar` from `Foo::Bar`.
+                            new_bound.segments.pop();
+                            // Add `<Bar = RhsTy>` to `Foo` if applicable and return.
+                            match &mut new_bound.segments.last_mut().unwrap().args {
+                                Some(args) => match args.deref_mut() {
+                                    GenericArgs::Parenthesized(_) => None,
+                                    GenericArgs::AngleBracketed(args) => {
+                                        args.args.push(new_arg);
+                                        Some(new_bound)
+                                    }
+                                },
+                                empty_args => {
+                                    *empty_args = AngleBracketedArgs {
+                                        span: assoc_ty_ident.span,
+                                        args: vec![new_arg],
                                     }
-                                    err.span_suggestion_verbose(
+                                    .into();
+                                    Some(new_bound)
+                                }
+                            }
+                        };
+
+                        if let Some(bound) = bound_opt {
+                            err.assoc_constraint_suggestion.push(
+                                EqualityConstraintToAssocConstraintSuggestion {
+                                    assoc_ty: assoc_ty_ident.name.to_string(),
```

Comment:
I think you just use the `Symbol` directly in the type rather than calling `.to_string()` on it (same elsewhere).

Reply:
Thanks, there was a bug. Storing `Symbol` or calling `Symbol.to_string()` doesn't work since we have to deal with `r#` identifiers, especially in code suggestions. I should have stored either `Ident` or `Ident.to_string()`.
There's another case of EqualityConstraintToAssocConstraintSuggestion, whose assoc_ty is not a single identifier but a concatenation. To support that case I would have to store String or make that suggestion an enum. I want to choose the first option for the sake of simplicity.